File: src/lib/git.ts
import { invoke } from "@tauri-apps/api/core";

export interface Commit {
  hash: string;
  short_hash: string;
  author: string;
  age: string;
  subject: string;
  refs: string[];
  parents: string[];
}

export interface FileEntry {
  path: string;
  status: string;
}

export interface CommitFile {
  path: string;
  status: string;
}

export interface WorkingStatus {
  branch: string;
  ahead: number;
  behind: number;
  staged: FileEntry[];
  unstaged: FileEntry[];
  untracked: string[];
}

export async function gitHistory(
  projectRoot: string,
  limit = 100,
  offset = 0,
): Promise<Commit[]> {
  return invoke<Commit[]>("git_history", { projectRoot, limit, offset });
}

export async function gitShow(
  projectRoot: string,
  hash: string,
): Promise<string> {
  return invoke<string>("git_show", { projectRoot, hash });
}

/** Files changed in a single commit (status + path). */
export async function gitCommitFiles(
  projectRoot: string,
  hash: string,
): Promise<CommitFile[]> {
  return invoke<CommitFile[]>("git_commit_files", { projectRoot, hash });
}

/** Unified diff for one file within a commit. */
export async function gitCommitFileDiff(
  projectRoot: string,
  hash: string,
  path: string,
): Promise<string> {
  return invoke<string>("git_commit_file_diff", { projectRoot, hash, path });
}

export async function gitWorkingStatus(
  projectRoot: string,
): Promise<WorkingStatus> {
  return invoke<WorkingStatus>("git_working_status", { projectRoot });
}

export async function gitStageFile(
  projectRoot: string,
  path: string,
): Promise<void> {
  return invoke("git_stage_file", { projectRoot, path });
}

export async function gitUnstageFile(
  projectRoot: string,
  path: string,
): Promise<void> {
  return invoke("git_unstage_file", { projectRoot, path });
}

export async function gitDiscardChanges(
  projectRoot: string,
  path: string,
): Promise<void> {
  return invoke("git_discard_changes", { projectRoot, path });
}

/** Which side of the index to diff for `gitFileDiff`. */
export type DiffMode = "staged" | "unstaged" | "untracked";

/**
 * Unified diff text for one file. `staged` compares index↔HEAD, `unstaged`
 * compares working-tree↔index, and `untracked` synthesizes an all-additions
 * patch so new files are reviewable too. Output is capped server-side.
 */
export async function gitFileDiff(
  projectRoot: string,
  path: string,
  mode: DiffMode,
): Promise<string> {
  return invoke<string>("git_file_diff", { projectRoot, path, mode });
}

export async function gitCommit(
  projectRoot: string,
  message: string,
): Promise<void> {
  return invoke("git_commit", { projectRoot, message });
}

File: src-tauri/src/commands/git.rs
//! Tauri command wrappers around [`crate::git`].
//!
//! Each handler is a thin async shim that converts the request payload into a
//! `Path` + calls into the sync implementation. Returning `Result<_, String>`
//! matches the existing command-bus convention.

use std::path::PathBuf;

use crate::git::{self, Commit, CommitFile, WorkingStatus};

#[tauri::command]
pub async fn git_history(
    project_root: String,
    limit: u32,
    offset: Option<u32>,
) -> Result<Vec<Commit>, String> {
    let root = PathBuf::from(&project_root);
    git::history(&root, limit, offset.unwrap_or(0))
}

#[tauri::command]
pub async fn git_show(project_root: String, hash: String) -> Result<String, String> {
    let root = PathBuf::from(&project_root);
    git::show_commit(&root, &hash)
}

#[tauri::command]
pub async fn git_commit_files(
    project_root: String,
    hash: String,
) -> Result<Vec<CommitFile>, String> {
    let root = PathBuf::from(&project_root);
    git::commit_files(&root, &hash)
}

#[tauri::command]
pub async fn git_commit_file_diff(
    project_root: String,
    hash: String,
    path: String,
) -> Result<String, String> {
    let root = PathBuf::from(&project_root);
    git::commit_file_diff(&root, &hash, &path)
}

#[tauri::command]
pub async fn git_working_status(project_root: String) -> Result<WorkingStatus, String> {
    let root = PathBuf::from(&project_root);
    git::working_status(&root)
}

#[tauri::command]
pub async fn git_stage_file(project_root: String, path: String) -> Result<(), String> {
    let root = PathBuf::from(&project_root);
    git::stage_file(&root, &path)
}

#[tauri::command]
pub async fn git_unstage_file(project_root: String, path: String) -> Result<(), String> {
    let root = PathBuf::from(&project_root);
    git::unstage_file(&root, &path)
}

#[tauri::command]
pub async fn git_discard_changes(project_root: String, path: String) -> Result<(), String> {
    let root = PathBuf::from(&project_root);
    git::discard_changes(&root, &path)
}

#[tauri::command]
pub async fn git_commit(project_root: String, message: String) -> Result<(), String> {
    let root = PathBuf::from(&project_root);
    git::commit_staged(&root, &message)
}

/// Unified diff for one file. `mode` is `"staged"`, `"unstaged"`, or
/// `"untracked"` (the latter synthesizes an all-additions patch).
#[tauri::command]
pub async fn git_file_diff(
    project_root: String,
    path: String,
    mode: String,
) -> Result<String, String> {
    let root = PathBuf::from(&project_root);
    let mode = git::DiffMode::parse(&mode)?;
    git::file_diff(&root, &path, mode)
}
